core/tls-websocket-server/src/ws_server.rs

```diff
@@ -18,71 +18,427 @@
+		let server_join_handle = thread::spawn(move || server_clone.run());
+
+		// Wait until server is up.
+		thread::sleep(std::time::Duration::from_millis(50));
+
+		let number_of_connections = 6usize;
+
+		// Spawn multiple clients that connect to the server simultaneously and send a message.
+		let client_handles: Vec<_> = (0..number_of_connections)
+			.map(|_| {
+				let expected_answer_clone = expected_answer.clone();
+
+				thread::spawn(move || {
+					let mut socket = connect_tls_client(get_server_addr(port).as_str());
+					socket
+						.write_message(Message::Text("Hello WebSocket".into()))
+						.expect("client write message to be successful");
+
+					assert_eq!(
+						Message::Text(expected_answer_clone),
+						socket.read_message().unwrap()
+					);
+				})
+			})
+			.collect();
+
+		for handle in client_handles.into_iter() {
+			handle.join().expect("client handle to be joined");
+		}
+
+		server.shut_down().unwrap();
+
+		let server_shutdown_result =
+			server_join_handle.join().expect("Couldn't join on the associated thread");
+		if let Err(e) = server_shutdown_result {
+			panic!("Test failed, web-socket returned error: {:?}", e);
+		}
+
+		assert_eq!(number_of_connections, handler.get_handled_messages().len());
+	}
+
+	#[test]
+	fn server_sends_update_message_to_client() {
+		let _ = env_logger::builder().is_test(true).try_init();
+
+		let expected_answer = "first response".to_string();
+		let port: u16 = 21778;
+		let (server, handler) = create_server(Some(expected_answer.clone()), port);
+
+		let server_clone = server.clone();
+		let server_join_handle = thread::spawn(move || server_clone.run());
 
-		let listener = TcpListener::bind(&socket_addr).map_err(WebSocketError::TcpBindError)?;
+		// Wait until server is up.
+		thread::sleep(std::time::Duration::from_millis(50));
 
+		let update_message = "Message update".to_string();
+		let update_message_clone = update_message.clone();
+
+		let client_join_handle = thread::spawn(move || {
+			let mut socket = connect_tls_client(get_server_addr(port).as_str());
+			socket
+				.write_message(Message::Text("First request".into()))
+				.expect("client write message to be successful");
+
+			assert_eq!(Message::Text(expected_answer), socket.read_message().unwrap());
+			assert_eq!(Message::Text(update_message_clone), socket.read_message().unwrap());
+		});
+
+		let connection_token = poll_handler_for_first_connection(handler.as_ref());
+
+		// Send reply to a wrong connection token, should fail.
+		assert!(server
+			.send_message(
+				ConnectionToken(connection_token.0 + 1),
+				"wont get to the client".to_string()
+			)
+			.is_err());
+
+		// Send reply to the correct connection token.
+		server.send_message(connection_token, update_message).unwrap();
+
+		client_join_handle.join().unwrap();
+		server.shut_down().unwrap();
+		server_join_handle.join().unwrap().unwrap();
+
+		assert_eq!(1, handler.get_handled_messages().len());
+	}
+
+	#[test]
+	#[ignore]
```

Comment:
why the ignore?

Reply:
Because it connects to an outside resource (`ws.ifelse.io`) and was mainly used for me to test the tungstenite web-socket client and its behavior with different certificates. It does not test any of our own components. However, IMO it can still be useful to run, in case our real tests break, and we want to verify it's not because of the client and certificates.